timeseries/src/autogluon/timeseries/models/chronos/model.py

```diff
@@ -0,0 +1,366 @@
+import logging
+import os
+from itertools import chain
+from pathlib import Path
+from typing import Any, Dict, Literal, Optional, Union
+
+import numpy as np
+import pandas as pd
+from transformers import AutoConfig, AutoModelForCausalLM, AutoModelForSeq2SeqLM
+
+from autogluon.common.loaders import load_pkl
+from autogluon.timeseries.dataset.ts_dataframe import ITEMID, TIMESTAMP, TimeSeriesDataFrame
+from autogluon.timeseries.models.abstract import AbstractTimeSeriesModel
+from autogluon.timeseries.utils.forecast import get_forecast_horizon_index_ts_dataframe
+
+from .chronos import ChronosConfig, ChronosPipeline, ChronosPretrainedModel
+
+# NOTE: We avoid imports for torch and lightning.pytorch at the top level and hide them inside class methods.
+# This is done to skip these imports during multiprocessing (which may cause bugs)
+
+logger = logging.getLogger(__name__)
+
+
+# allowed HuggingFace model paths with custom parameter definitions
+MODEL_CONFIGS = {
+    "amazon/chronos-t5-tiny": {
+        "num_gpus": 0,  # minimum number of required GPUs
+    },
+    "amazon/chronos-t5-mini": {"num_gpus": 0},
+    "amazon/chronos-t5-small": {"num_gpus": 1},
+    "amazon/chronos-t5-base": {"num_gpus": 1},
+    "amazon/chronos-t5-large": {"num_gpus": 1},
+}
+
+
+class ChronosInferenceDataset:
+    """A container for time series datasets that implements the ``torch.utils.data.Dataset`` interface"""
+
+    def __init__(
+        self,
+        target_df: TimeSeriesDataFrame,
+        context_length: int,
+        target_column: str = "target",
+    ):
+        assert target_df is not None
+        assert target_df.freq, "Initializing GluonTS data sets without freq is not allowed"
+
+        self.target_column = target_column
+        self.context_length = context_length
+        self.target_array = target_df[target_column].to_numpy(dtype=np.float32)
+        self.freq = target_df.freq
+        self._set_indptr(target_df=target_df)
+
+    def _set_indptr(self, target_df: TimeSeriesDataFrame):
+        """Replace inefficient groupby ITEMID with indptr that stores start:end of each time series"""
+
+        # TODO: refactor into generic iterator
+
+        item_id_index = target_df.index.get_level_values(ITEMID)
+        indices_sizes = item_id_index.value_counts(sort=False)
+        self.item_ids = indices_sizes.index  # shape [num_items]
+        cum_sizes = indices_sizes.values.cumsum()
+        self.indptr = np.append(0, cum_sizes).astype(np.int32)
+        self.start_timestamps = target_df.reset_index(TIMESTAMP).groupby(level=ITEMID, sort=False).first()[TIMESTAMP]
```

Comment:
Do we need to store/return `item_ids` and `start_timestamps`?

Reply:
why not? (i.e., I would like to keep `get_full_item` method as well.

Reply:
My reasoning was that 1/ Chronos won't use timestamp/item_id information until at least AG v1.2 and 2/ even when it does, it won't necessarily expect the GluonTS dataset format with start+freq fields (e.g., passing the timestamps as `pd.DatetimeIndex` might be more efficient). Therefore, these variables as well as `get_full_item` method would be dead code until v1.2. If we need this logic in the future, we can copy it from `SimpleGluonTSDataset`.